README.md
# Email with Mailgun

The new Email with Mailgun step.

Send emails with Mailgun [https://mailgun.com](https://mailgun.com)

Can be run directly with the [bitrise CLI](https://github.com/bitrise-io/bitrise),
just `git clone` this repository, `cd` into it's folder in your Terminal/Command Line
and call `bitrise run test`.

*Check the `bitrise.yml` file for required inputs which have to be
added to your `.bitrise.secrets.yml` file!*


CHANGELOG.md
## Changelog (Current version: 2.1.3)

-----------------

### 2.1.3 (2018 Apr 12)

* [bc3c4c7] Prepare for 2.1.3

### 2.1.2 (2018 Apr 10)

* [1ca1e4d] Prepare for version 2.1.2
* [2caebd9] Update step.yml (#19)

### 2.1.1 (2018 Jan 08)

* [a0dd83e] Bump version to 2.1.1
* [f23d3bd] Add note what to do if you don't get the email (#18)
* [8503147] Merge pull request #17 from bitrise-io/fix/multiple-attachments
* [58c4053] Minor fixes
* [aebe710] Fix package name
* [1025bf0] remove comment
* [b00c2d7] Set go toolkit
* [0b26553] Upload all attachments not just the last one

### 2.1.0 (2016 Apr 12)

* [69cfa00] Merge pull request #14 from bitrise-io/step_revision
* [fa1ef05] cleanup
* [07bf97d] step.yml fix
* [2291eef] step revision
* [e8e27a0] Merge pull request #13 from mackoj/master
* [8d74905] Support for file attachments
* [8b3428d] Merge pull request #10 from bazscsa/patch-1
* [7d6aa31] Update step.yml

### 2.0.0 (2015 Sep 08)

* [10dd6e4] bitrise stack related update
* [53ee940] Merge pull request #8 from gkiki90/update
* [a76a8dd] gitignore
* [ffccc05] update
* [8837365] Update step_test.sh
* [7f99e1f] Update step.yml
* [da20e9e] step.yml update
* [2f82866] Merge pull request #3 from erosdome/master
* [9339252] Update step.yml
* [48a95c0] Update step.yml
* [7e324aa] Update README.md

### 1.1.0 (2014 Sep 08)

* [35c62ae] description link markdown
* [efbf49c] Concrete->Bitrise text changes
* [32c0e47] Merge pull request #2 from erosdome/master
* [e0cc478] initial commit
* [c381e80] Merge pull request #1 from tomfurrier/master
* [270f0a1] removed unnecessary test
* [08a47a3] input env var validation added
* [ffb4e08] output formatting, env var scope fixes in tests
* [2760d26] env var values fixed in tests
* [9e43a19] added parameter tests
* [f7c2ab4] added tests, modified read me

### 1.0.2 (2014 Jun 12)

* [8dcb6f5] Merge branch 'release/1.0.2'
* [9bf117b] step.yml fix

-----------------

Updated: 2018 Apr 12

step_test.sh
#!/bin/bash

#
# Run it from the directory which contains step.sh
#


# ------------------------
# --- Helper functions ---

function print_and_do_command {
  echo "$ $@"
  $@
}

function inspect_test_result {
  if [ $1 -eq 0 ]; then
    test_results_success_count=$[test_results_success_count + 1]
  else
    test_results_error_count=$[test_results_error_count + 1]
  fi
}

#
# First param is the expect message, other are the command which will be executed.
#
function expect_success {
  expect_msg=$1
  shift

  echo " -> $expect_msg"
  $@
  cmd_res=$?

  if [ $cmd_res -eq 0 ]; then
    echo " [OK] Expected zero return code, got: 0"
  else
    echo " [ERROR] Expected zero return code, got: $cmd_res"
    exit 1
  fi
}

#
# First param is the expect message, other are the command which will be executed.
#
function expect_error {
  expect_msg=$1
  shift

  echo " -> $expect_msg"
  $@
  cmd_res=$?

  if [ ! $cmd_res -eq 0 ]; then
    echo " [OK] Expected non-zero return code, got: $cmd_res"
  else
    echo " [ERROR] Expected non-zero return code, got: 0"
    exit 1
  fi
}

function is_dir_exist {
  if [ -d "$1" ]; then
    return 0
  else
    return 1
  fi
}

function is_file_exist {
  if [ -f "$1" ]; then
    return 0
  else
    return 1
  fi
}

function is_not_unset_or_empty {
  if [[ $1 ]]; then
    return 0
  else
    return 1
  fi
}

function test_env_cleanup {
  unset api_key
	unset domain
	unset send_to
	unset subject
	unset message
}

function print_new_test {
  echo
  echo "[TEST]"
}

# -----------------
# --- Run tests ---


echo "Starting tests..."

test_results_success_count=0
test_results_error_count=0


# [TEST] Call the command with api_key not set,
# it should raise an error message and exit
#
(
  print_new_test
	test_env_cleanup

  # Set env var
  export domain="dsa4321"
  export send_to="some@email.com"
  export subject="Bitrise Email Test"
  export message="It works from tests!"

   # All of the required env vars should exist except api_key
  expect_error "api_key environment variable should NOT be set" is_not_unset_or_empty "$api_key"
  expect_success "domain environment variable should be set" is_not_unset_or_empty "$domain"
  expect_success "send_to environment variable should be set" is_not_unset_or_empty "$send_to"
  expect_success "subject environment variable should be set" is_not_unset_or_empty "$subject"
	expect_success "message environment variable should be set" is_not_unset_or_empty "$message"

  # Send email request
  expect_error "The command should be called, but should not complete sucessfully" print_and_do_command ./step.sh
)
test_result=$?
inspect_test_result $test_result


# [TEST] Call the command with domain not set,
# it should raise an error message and exit
#
(
  print_new_test
	test_env_cleanup

  # Set env var
  export api_key="asd1234"
  export send_to="some@email.com"
  export subject="Bitrise Email Test"
  export message="It works from tests!"

   # All of the required env vars should exist except domain
  expect_success "api_key environment variable should be set" is_not_unset_or_empty "$api_key"
  expect_error "domain environment variable should NOT be set" is_not_unset_or_empty "$domain"
  expect_success "send_to environment variable should be set" is_not_unset_or_empty "$send_to"
  expect_success "subject environment variable should be set" is_not_unset_or_empty "$subject"
	expect_success "message environment variable should be set" is_not_unset_or_empty "$message"

  # Send email request
  expect_error "The command should be called, but should not complete sucessfully" print_and_do_command ./step.sh
)
test_result=$?
inspect_test_result $test_result


# [TEST] Call the command with send_to not set,
# it should raise an error message and exit
#
(
  print_new_test
	test_env_cleanup

  # Set env var
  export api_key="asd1234"
  export domain="dsa4321"
  export subject="Bitrise Email Test"
  export message="It works from tests!"

   # All of the required env vars should exist except send_to
  expect_success "api_key environment variable should be set" is_not_unset_or_empty "$api_key"
  expect_success "domain environment variable should be set" is_not_unset_or_empty "$domain"
  expect_error "send_to environment variable should NOT be set" is_not_unset_or_empty "$send_to"
  expect_success "subject environment variable should be set" is_not_unset_or_empty "$subject"
	expect_success "message environment variable should be set" is_not_unset_or_empty "$message"

  # Send email request
  expect_error "The command should be called, but should not complete sucessfully" print_and_do_command ./step.sh
)
test_result=$?
inspect_test_result $test_result


# an email can be sent without subject but it is not recommended


# [TEST] Call the command with message not set,
# it should raise an error message and exit
#
(
  print_new_test
  test_env_cleanup

  # Set env var
  export api_key="asd1234"
  export domain="dsa4321"
  export send_to="asd1234"
  export subject="Bitrise Email Test"

   # All of the required env vars should exist except send_to
  expect_success "api_key environment variable should be set" is_not_unset_or_empty "$api_key"
  expect_success "domain environment variable should be set" is_not_unset_or_empty "$domain"
  expect_success "send_to environment variable should be set" is_not_unset_or_empty "$send_to"
  expect_success "subject environment variable should be set" is_not_unset_or_empty "$subject"
  expect_error "message environment variable should NOT be set" is_not_unset_or_empty "$message"

  # Send email request
  expect_error "The command should be called, but should not complete sucessfully" print_and_do_command ./step.sh
)
test_result=$?
inspect_test_result $test_result


#final cleanup
test_env_cleanup

# --------------------
# --- Test Results ---

echo
echo "--- Results ---"
echo " * Errors: $test_results_error_count"
echo " * Success: $test_results_success_count"
echo "---------------"

if [ $test_results_error_count -eq 0 ]; then
  echo "-> SUCCESS"
else
  echo "-> FAILED"
fi


tests/step_test.sh
#!/bin/bash

#
# Run it from the directory which contains step.sh
#


# ------------------------
# --- Helper functions ---

function print_and_do_command {
  echo "$ $@"
  $@
}

function inspect_test_result {
  if [ $1 -eq 0 ]; then
    test_results_success_count=$[test_results_success_count + 1]
  else
    test_results_error_count=$[test_results_error_count + 1]
  fi
}

#
# First param is the expect message, other are the command which will be executed.
#
function expect_success {
  expect_msg=$1
  shift

  echo " -> $expect_msg"
  $@
  cmd_res=$?

  if [ $cmd_res -eq 0 ]; then
    echo " [OK] Expected zero return code, got: 0"
  else
    echo " [ERROR] Expected zero return code, got: $cmd_res"
    exit 1
  fi
}

#
# First param is the expect message, other are the command which will be executed.
#
function expect_error {
  expect_msg=$1
  shift

  echo " -> $expect_msg"
  $@
  cmd_res=$?

  if [ ! $cmd_res -eq 0 ]; then
    echo " [OK] Expected non-zero return code, got: $cmd_res"
  else
    echo " [ERROR] Expected non-zero return code, got: 0"
    exit 1
  fi
}

function is_dir_exist {
  if [ -d "$1" ]; then
    return 0
  else
    return 1
  fi
}

function is_file_exist {
  if [ -f "$1" ]; then
    return 0
  else
    return 1
  fi
}

function is_not_unset_or_empty {
  if [[ $1 ]]; then
    return 0
  else
    return 1
  fi
}

function test_env_cleanup {
  unset MAILGUN_API_KEY
	unset MAILGUN_DOMAIN
	unset MAILGUN_SEND_TO
	unset MAILGUN_EMAIL_SUBJECT
	unset MAILGUN_EMAIL_MESSAGE
}

function print_new_test {
  echo
  echo "[TEST]"
}

# -----------------
# --- Run tests ---


echo "Starting tests..."

test_results_success_count=0
test_results_error_count=0


# [TEST] Call the command with MAILGUN_API_KEY not set, 
# it should raise an error message and exit
# 
(
  print_new_test
	test_env_cleanup

  # Set env var
  export MAILGUN_DOMAIN="dsa4321"
  export MAILGUN_SEND_TO="some@email.com"
  export MAILGUN_EMAIL_SUBJECT="Bitrise Email Test"
  export MAILGUN_EMAIL_MESSAGE="It works from tests!"

   # All of the required env vars should exist except MAILGUN_API_KEY
  expect_error "MAILGUN_API_KEY environment variable should NOT be set" is_not_unset_or_empty "$MAILGUN_API_KEY"
  expect_success "MAILGUN_DOMAIN environment variable should be set" is_not_unset_or_empty "$MAILGUN_DOMAIN"
  expect_success "MAILGUN_SEND_TO environment variable should be set" is_not_unset_or_empty "$MAILGUN_SEND_TO"
  expect_success "MAILGUN_EMAIL_SUBJECT environment variable should be set" is_not_unset_or_empty "$MAILGUN_EMAIL_SUBJECT"
	expect_success "MAILGUN_EMAIL_MESSAGE environment variable should be set" is_not_unset_or_empty "$MAILGUN_EMAIL_MESSAGE"

  # Send email request
  expect_error "The command should be called, but should not complete sucessfully" print_and_do_command ./step.sh
)
test_result=$?
inspect_test_result $test_result


# [TEST] Call the command with MAILGUN_DOMAIN not set, 
# it should raise an error message and exit
# 
(
  print_new_test
	test_env_cleanup

  # Set env var
  export MAILGUN_API_KEY="asd1234"
  export MAILGUN_SEND_TO="some@email.com"
  export MAILGUN_EMAIL_SUBJECT="Bitrise Email Test"
  export MAILGUN_EMAIL_MESSAGE="It works from tests!"

   # All of the required env vars should exist except MAILGUN_DOMAIN
  expect_success "MAILGUN_API_KEY environment variable should be set" is_not_unset_or_empty "$MAILGUN_API_KEY"
  expect_error "MAILGUN_DOMAIN environment variable should NOT be set" is_not_unset_or_empty "$MAILGUN_DOMAIN"
  expect_success "MAILGUN_SEND_TO environment variable should be set" is_not_unset_or_empty "$MAILGUN_SEND_TO"
  expect_success "MAILGUN_EMAIL_SUBJECT environment variable should be set" is_not_unset_or_empty "$MAILGUN_EMAIL_SUBJECT"
	expect_success "MAILGUN_EMAIL_MESSAGE environment variable should be set" is_not_unset_or_empty "$MAILGUN_EMAIL_MESSAGE"

  # Send email request
  expect_error "The command should be called, but should not complete sucessfully" print_and_do_command ./step.sh
)
test_result=$?
inspect_test_result $test_result


# [TEST] Call the command with MAILGUN_SEND_TO not set, 
# it should raise an error message and exit
# 
(
  print_new_test
	test_env_cleanup

  # Set env var
  export MAILGUN_API_KEY="asd1234"
  export MAILGUN_DOMAIN="dsa4321"
  export MAILGUN_EMAIL_SUBJECT="Bitrise Email Test"
  export MAILGUN_EMAIL_MESSAGE="It works from tests!"

   # All of the required env vars should exist except MAILGUN_SEND_TO
  expect_success "MAILGUN_API_KEY environment variable should be set" is_not_unset_or_empty "$MAILGUN_API_KEY"
  expect_success "MAILGUN_DOMAIN environment variable should be set" is_not_unset_or_empty "$MAILGUN_DOMAIN"
  expect_error "MAILGUN_SEND_TO environment variable should NOT be set" is_not_unset_or_empty "$MAILGUN_SEND_TO"
  expect_success "MAILGUN_EMAIL_SUBJECT environment variable should be set" is_not_unset_or_empty "$MAILGUN_EMAIL_SUBJECT"
	expect_success "MAILGUN_EMAIL_MESSAGE environment variable should be set" is_not_unset_or_empty "$MAILGUN_EMAIL_MESSAGE"

  # Send email request
  expect_error "The command should be called, but should not complete sucessfully" print_and_do_command ./step.sh
)
test_result=$?
inspect_test_result $test_result


# an email can be sent without subject but it is not recommended


# [TEST] Call the command with MAILGUN_EMAIL_MESSAGE not set, 
# it should raise an error message and exit
# 
(
  print_new_test
  test_env_cleanup

  # Set env var
  export MAILGUN_API_KEY="asd1234"
  export MAILGUN_DOMAIN="dsa4321"
  export MAILGUN_SEND_TO="asd1234"
  export MAILGUN_EMAIL_SUBJECT="Bitrise Email Test"

   # All of the required env vars should exist except MAILGUN_SEND_TO
  expect_success "MAILGUN_API_KEY environment variable should be set" is_not_unset_or_empty "$MAILGUN_API_KEY"
  expect_success "MAILGUN_DOMAIN environment variable should be set" is_not_unset_or_empty "$MAILGUN_DOMAIN"
  expect_success "MAILGUN_SEND_TO environment variable should be set" is_not_unset_or_empty "$MAILGUN_SEND_TO"
  expect_success "MAILGUN_EMAIL_SUBJECT environment variable should be set" is_not_unset_or_empty "$MAILGUN_EMAIL_SUBJECT"
  expect_error "MAILGUN_EMAIL_MESSAGE environment variable should NOT be set" is_not_unset_or_empty "$MAILGUN_EMAIL_MESSAGE"

  # Send email request
  expect_error "The command should be called, but should not complete sucessfully" print_and_do_command ./step.sh
)
test_result=$?
inspect_test_result $test_result


#final cleanup
test_env_cleanup

# --------------------
# --- Test Results ---

echo
echo "--- Results ---"
echo " * Errors: $test_results_error_count"
echo " * Success: $test_results_success_count"
echo "---------------"

if [ $test_results_error_count -eq 0 ]; then
  echo "-> SUCCESS"
else
  echo "-> FAILED"
fi


step.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"mime/multipart"
	"net/http"
	"os"
	"strings"
)

// -----------------------
// --- Functions
// -----------------------

func logFail(format string, v ...interface{}) {
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("\x1b[31;1m%s\x1b[0m\n", errorMsg)
	os.Exit(1)
}

func logWarn(format string, v ...interface{}) {
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("\x1b[33;1m%s\x1b[0m\n", errorMsg)
}

func logInfo(format string, v ...interface{}) {
	fmt.Println()
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("\x1b[34;1m%s\x1b[0m\n", errorMsg)
}

func logDetails(format string, v ...interface{}) {
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("  %s\n", errorMsg)
}

func logDone(format string, v ...interface{}) {
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("  \x1b[32;1m%s\x1b[0m\n", errorMsg)
}

func validateRequiredInput(key string) string {
	value := os.Getenv(key)
	if value == "" {
		logFail("missing required input: %s", key)
	}
	return value
}

func strip(str string) string {
	dirty := true
	strippedStr := str
	for dirty {
		hasWhiteSpacePrefix := false
		if strings.HasPrefix(strippedStr, " ") {
			hasWhiteSpacePrefix = true
			strippedStr = strings.TrimPrefix(strippedStr, " ")
		}

		hasWhiteSpaceSuffix := false
		if strings.HasSuffix(strippedStr, " ") {
			hasWhiteSpaceSuffix = true
			strippedStr = strings.TrimSuffix(strippedStr, " ")
		}

		if !hasWhiteSpacePrefix && !hasWhiteSpaceSuffix {
			dirty = false
		}
	}
	return strippedStr
}

func genericIsPathExists(pth string) (os.FileInfo, bool, error) {
	if pth == "" {
		return nil, false, errors.New("No path provided")
	}
	fileInf, err := os.Stat(pth)
	if err == nil {
		return fileInf, true, nil
	}
	if os.IsNotExist(err) {
		return nil, false, nil
	}
	return fileInf, false, err
}

func isPathExists(pth string) (bool, error) {
	_, isExists, err := genericIsPathExists(pth)
	return isExists, err
}

func createRequest(url string, fields map[string]string, attachments []string) (*http.Request, error) {
	var b bytes.Buffer
	w := multipart.NewWriter(&b)

	// Add fields
	for key, value := range fields {
		if err := w.WriteField(key, value); err != nil {
			return nil, err
		}
	}

	// Add files
	for _, file := range attachments {
		f, err := os.Open(file)
		if err != nil {
			return nil, err
		}
		fw, err := w.CreateFormFile("attachment", file)
		if err != nil {
			return nil, err
		}
		if _, err = io.Copy(fw, f); err != nil {
			return nil, err
		}
	}

	w.Close()

	req, err := http.NewRequest("POST", url, &b)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", w.FormDataContentType())

	return req, nil
}

// -----------------------
// --- Main
// -----------------------

func main() {
	//
	// Validate options
	logInfo("Configs:")
	logDetails("api_key: %s", os.Getenv("api_key"))
	logDetails("domain: %s", os.Getenv("domain"))
	logDetails("from_email: %s", os.Getenv("from_email"))
	logDetails("send_to: %s", os.Getenv("send_to"))
	logDetails("subject: %s", os.Getenv("subject"))
	logDetails("message: %s", os.Getenv("message"))
	logDetails("error_message: %s", os.Getenv("error_message"))
	logDetails("message_format: %s", os.Getenv("message_format"))
	logDetails("attachments: %s", os.Getenv("attachments"))

	apiKey := validateRequiredInput("api_key")
	domain := validateRequiredInput("domain")
	fromEmail := validateRequiredInput("from_email")
	sendTo := validateRequiredInput("send_to")
	subject := validateRequiredInput("subject")
	message := validateRequiredInput("message")
	messageFormat := validateRequiredInput("message_format")

	errorMessage := os.Getenv("error_message")
	attachmentList := os.Getenv("attachments")

	isBuildFailedMode := (os.Getenv("STEPLIB_BUILD_STATUS") != "0")
	if isBuildFailedMode && errorMessage != "" {
		message = errorMessage
	}

	attachments := []string{}
	if attachmentList != "" {
		split := strings.Split(attachmentList, ",")
		for _, item := range split {
			filePath := strip(item)
			if exist, err := isPathExists(filePath); err != nil {
				logFail("failed to check if path (%s) exist, error: %s", filePath, err)
			} else if !exist {
				logFail("file not exist at (%s)", filePath)
			}

			attachments = append(attachments, filePath)
		}
	}

	//
	// Create request
	logInfo("Performing request")

	requestURL := fmt.Sprintf("https://api.mailgun.net/v2/%s/messages", domain)

	fields := map[string]string{
		"from":    fromEmail,
		"to":      sendTo,
		"subject": subject,
	}

	switch messageFormat {
	case "html":
		fields["html"] = message
	case "text":
		fields["text"] = message
	default:
		logFail("invalid message_format: %s, available options [html, text]", messageFormat)
	}

	request, err := createRequest(requestURL, fields, attachments)
	if err != nil {
		logFail("Failed to create request, error: %#v", err)
	}
	request.SetBasicAuth("api", apiKey)

	client := http.Client{}
	response, requestErr := client.Do(request)
	if requestErr != nil {
		logFail("Performing request failed, error: %#v", requestErr)
	}

	defer response.Body.Close()
	contents, readErr := ioutil.ReadAll(response.Body)
	if readErr != nil {
		logWarn("Failed to read response body, error: %#v", readErr)
	}

	logInfo("response content: %s", contents)
	logDetails("status code: %d", response.StatusCode)
}


main.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"mime/multipart"
	"net/http"
	"os"
	"strings"
)

// -----------------------
// --- Functions
// -----------------------

func logFail(format string, v ...interface{}) {
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("\x1b[31;1m%s\x1b[0m\n", errorMsg)
	os.Exit(1)
}

func logWarn(format string, v ...interface{}) {
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("\x1b[33;1m%s\x1b[0m\n", errorMsg)
}

func logInfo(format string, v ...interface{}) {
	fmt.Println()
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("\x1b[34;1m%s\x1b[0m\n", errorMsg)
}

func logDetails(format string, v ...interface{}) {
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("  %s\n", errorMsg)
}

func logDone(format string, v ...interface{}) {
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("  \x1b[32;1m%s\x1b[0m\n", errorMsg)
}

func validateRequiredInput(key string) string {
	value := os.Getenv(key)
	if value == "" {
		logFail("missing required input: %s", key)
	}
	return value
}

func strip(str string) string {
	dirty := true
	strippedStr := str
	for dirty {
		hasWhiteSpacePrefix := false
		if strings.HasPrefix(strippedStr, " ") {
			hasWhiteSpacePrefix = true
			strippedStr = strings.TrimPrefix(strippedStr, " ")
		}

		hasWhiteSpaceSuffix := false
		if strings.HasSuffix(strippedStr, " ") {
			hasWhiteSpaceSuffix = true
			strippedStr = strings.TrimSuffix(strippedStr, " ")
		}

		if !hasWhiteSpacePrefix && !hasWhiteSpaceSuffix {
			dirty = false
		}
	}
	return strippedStr
}

func genericIsPathExists(pth string) (os.FileInfo, bool, error) {
	if pth == "" {
		return nil, false, errors.New("No path provided")
	}
	fileInf, err := os.Stat(pth)
	if err == nil {
		return fileInf, true, nil
	}
	if os.IsNotExist(err) {
		return nil, false, nil
	}
	return fileInf, false, err
}

func isPathExists(pth string) (bool, error) {
	_, isExists, err := genericIsPathExists(pth)
	return isExists, err
}

func createRequest(url string, fields map[string]string, attachments []string) (*http.Request, error) {
	var b bytes.Buffer
	w := multipart.NewWriter(&b)

	// Add fields
	for key, value := range fields {
		if err := w.WriteField(key, value); err != nil {
			return nil, err
		}
	}

	// Add files
	for _, file := range attachments {
		f, err := os.Open(file)
		if err != nil {
			return nil, err
		}
		fw, err := w.CreateFormFile("attachment", file)
		if err != nil {
			return nil, err
		}
		if _, err = io.Copy(fw, f); err != nil {
			return nil, err
		}
	}

	if err := w.Close(); err != nil {
		return nil, err
	}

	req, err := http.NewRequest("POST", url, &b)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", w.FormDataContentType())

	return req, nil
}

// -----------------------
// --- Main
// -----------------------

func main() {
	//
	// Validate options
	logInfo("Configs:")
	logDetails("api_key: %s", os.Getenv("api_key"))
	logDetails("domain: %s", os.Getenv("domain"))
	logDetails("from_email: %s", os.Getenv("from_email"))
	logDetails("send_to: %s", os.Getenv("send_to"))
	logDetails("subject: %s", os.Getenv("subject"))
	logDetails("message: %s", os.Getenv("message"))
	logDetails("error_message: %s", os.Getenv("error_message"))
	logDetails("message_format: %s", os.Getenv("message_format"))
	logDetails("attachments: %s", os.Getenv("attachments"))

	apiKey := validateRequiredInput("api_key")
	domain := validateRequiredInput("domain")
	fromEmail := validateRequiredInput("from_email")
	sendTo := validateRequiredInput("send_to")
	subject := validateRequiredInput("subject")
	message := validateRequiredInput("message")
	messageFormat := validateRequiredInput("message_format")

	errorMessage := os.Getenv("error_message")
	attachmentList := os.Getenv("attachments")

	isBuildFailedMode := (os.Getenv("STEPLIB_BUILD_STATUS") != "0")
	if isBuildFailedMode && errorMessage != "" {
		message = errorMessage
	}

	attachments := []string{}
	if attachmentList != "" {
		split := strings.Split(attachmentList, ",")
		for _, item := range split {
			filePath := strip(item)
			if exist, err := isPathExists(filePath); err != nil {
				logFail("failed to check if path (%s) exist, error: %s", filePath, err)
			} else if !exist {
				logFail("file not exist at (%s)", filePath)
			}

			attachments = append(attachments, filePath)
		}
	}

	//
	// Create request
	logInfo("Performing request")

	requestURL := fmt.Sprintf("https://api.mailgun.net/v2/%s/messages", domain)

	fields := map[string]string{
		"from":    fromEmail,
		"to":      sendTo,
		"subject": subject,
	}

	switch messageFormat {
	case "html":
		fields["html"] = message
	case "text":
		fields["text"] = message
	default:
		logFail("invalid message_format: %s, available options [html, text]", messageFormat)
	}

	request, err := createRequest(requestURL, fields, attachments)
	if err != nil {
		logFail("Failed to create request, error: %s", err)
	}
	request.SetBasicAuth("api", apiKey)

	client := http.Client{}
	response, requestErr := client.Do(request)
	if requestErr != nil {
		logFail("Performing request failed, error: %s", requestErr)
	}

	defer func() {
		if err := response.Body.Close(); err != nil {
			logWarn("Failed to close response body, error: %s", err)
		}
	}()
	contents, readErr := ioutil.ReadAll(response.Body)
	if readErr != nil {
		logWarn("Failed to read response body, error: %s", readErr)
	}

	logInfo("response content: %s", contents)
	logDetails("status code: %d", response.StatusCode)
}
